### PR TITLE
Demo drivers

### PR DESCRIPTION
**Pull request by ranechita**

### CODE REVIEW

**Comment by ranechita** [comment]
v2

- ran astyle.sh

File: drivers/adc/adc_demo/adc_demo.h [hunk 1]
```diff
@@ -50,22 +50,34 @@
...
 };
 
 /**
  * @struct iio_demo_adc_init_param
  * @brief iio demo adc configuration.
  */
 struct adc_demo_init_param {
+	/* NUmber of channels*/
+	uint32_t channel_no;
+	/**buffer for communication between adc&dac, according to appropriate channel*/
```
Comment (mchindri) [inline drivers/adc/adc_demo/adc_demo.h:79]: I think this can be described as array of buffers for each channel.

File: drivers/adc/adc_demo/adc_demo.h [hunk 2]
```diff
@@ -50,22 +50,34 @@
...
+	uint16_t **loopback;
```
Comment (mchindri) [inline drivers/adc/adc_demo/adc_demo.h:80]: loopback_buffers maybe :)

File: projects/iio_demo/src/app/main.c [hunk 3]
```diff
@@ -194,12 +213,48 @@ int main(void)
...
+		.buff = (void *)ADC_DDR_BASEADDR,
+		.size = MAX_SIZE_BASE_ADDR
+	};
+
+	struct iio_data_buffer buff1 = {
+		.buff = (void *)DAC_DDR_BASEADDR,
+		.size = MAX_SIZE_BASE_ADDR
+	};
+
+	uint16_t** buf = (uint16_t**)malloc(10*sizeof(uint16_t*));
```
Comment (mchindri) [inline projects/iio_demo/src/app/main.c:226]: in iio_adc_demo.h : #define ADC_NUM_CHANNELS or something
and here: uint16_t loopback_buff[ADC_NUM_CHANNELS][MAX_SAMPLES_PER_CHANNEL]

File: projects/iio_demo/src/app/main.c [hunk 4]
```diff
@@ -194,12 +213,48 @@ int main(void)
 		.size = MAX_SIZE_BASE_ADDR
 	};
 
+	struct iio_data_buffer buff = {
```
Comment (mchindri) [inline projects/iio_demo/src/app/main.c:216]: adc_buff and dac_buff is a better naming

File: drivers/adc/adc_demo/iio_adc_demo.h [hunk 5]
```diff
@@ -42,10 +42,29 @@
...
-	.debug_reg_read = (int32_t (*)()) adc_demo_reg_read,
-	.debug_reg_write = (int32_t (*)()) adc_demo_reg_write
+#define MAX_NR_CHANNELS 10
+
+#define ADC_DEMO_DEV(_numch) {\
+	.num_ch = _numch, \
+	.attributes = NULL,	\
+	.debug_attributes = NULL,	\
+	.buffer_attributes = NULL,	\
+	.prepare_transfer = init_adc_channels,	\
```
Comment (mchindri) [inline drivers/adc/adc_demo/iio_adc_demo.h:54]: update_active_adc_channels should go here

File: drivers/adc/adc_demo/adc_demo.c [hunk 6]
```diff
@@ -85,6 +108,79 @@ int32_t adc_demo_remove(struct adc_demo_desc *desc)
 	return SUCCESS;
 }
 
+int32_t adc_get_number_of_channels(void* dev)
```
[the remark this answers is not in the document]
Comment (ranechita) [reply drivers/adc/adc_demo/adc_demo.c:111]: I forgot to delete it, it was from a previous approach. Removed it now

File: drivers/dac/dac_demo/dac_demo.h [hunk 7]
```diff
@@ -50,22 +50,32 @@
 /*************************** Types Declarations *******************************/
 /******************************************************************************/
 
-#define MAX_REG_ADDR		16
+#define MAX_DAC_ADDR		16
+#define DEFAULT_LOCAL_SAMPLES 128
+#define SINE_VECTOR_SIZE 128
```
[the remark this answers is not in the document]
Comment (ranechita) [reply drivers/dac/dac_demo/dac_demo.h:55]: removed it

File: drivers/adc/adc_demo/iio_adc_demo.h (hunk 5 above)
Comment (ranechita) [reply drivers/adc/adc_demo/iio_adc_demo.h:54]: the number of channels is now configurable from the init param, and that function retrieves the actual number of active channels

**Comment by ranechita** [comment]
v3

- init param field loopback now named loopback_buffers
- renamed iio_Data_buffers buff&buff1 to adc_buff and dac_buff
- the buffer from main `uint16_t** buf = (uint16_t**)malloc(10*sizeof(uint16_t*));` -> `uint16_t loopback_buff[MAX_ADC_CHANNELS][MAX_SAMPLES_PER_CHANNEL]`
- used array size instead of number `buff[k] = sine_lut[k%128];` -> `buff[k] = sine_lut[k % ARRAY_SIZE(sine_lut)]`
- removed unused function `int32_t adc_get_number_of_channels(void* dev)`
- removed unused define for sine_lut size

File: drivers/adc/adc_demo/adc_demo.h [hunk 8]
```diff
@@ -77,6 +89,14 @@ int32_t adc_demo_init(struct adc_demo_desc **desc,
 
 int32_t adc_demo_remove(struct adc_demo_desc *desc);
 
+int32_t adc_get_number_of_channels(void* dev);
```
[the remark this answers is not in the document]
Comment (ranechita) [reply drivers/adc/adc_demo/adc_demo.h]: was this a general remark (for functions like close_dac_channels,etc) or was it specific for this function?

File: drivers/adc/adc_demo/adc_demo.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (mchindri) [reply drivers/adc/adc_demo/adc_demo.c]: for all

File: drivers/adc/adc_demo/adc_demo.h (hunk 1 above)
Comment (ranechita) [reply drivers/adc/adc_demo/adc_demo.h:78]: Did you mean in the comment description?

File: drivers/adc/adc_demo/adc_demo.h [hunk 9]
```diff
@@ -50,22 +50,34 @@
...
 struct adc_demo_desc {
 	/** Dummy registers of device for testing */
-	uint8_t reg[MAX_REG_ADDR];
+	uint8_t reg[MAX_ADC_ADDR];
+	/** Active channel**/
+	uint32_t active_ch;
+	/** buffer for communication between adc&dac, according to channel*/
+	uint16_t **loopback;
+//	/* NUmber of channels*/
+//	uint32_t channel_no;
```
Comment (adrimbarean) [inline drivers/adc/adc_demo/adc_demo.h:68]: Is commented code needed?

Comment (ranechita) [reply drivers/adc/adc_demo/adc_demo.h:68]: i removed it

**Comment by ranechita** [comment]
v4

- removed unnecessary comment in `adc_demo_desc` in adc_demo.h
- shortened commit message

**Comment by ranechita** [comment]
v5

- moved `if(!dev) return -ENODEV;` after variable definition

File: drivers/adc/adc_demo/adc_demo.c [hunk 10]
```diff
@@ -85,6 +112,144 @@ int32_t adc_demo_remove(struct adc_demo_desc *desc)
...
+		*new_idx = -1;
+		return 0;
+	}
+	while (!(ch_mask & 1)) {
+		last_idx++;
+		ch_mask >>= 1;
+	}
+	*new_idx = last_idx;
+
+	return 1;
```
Comment (mirceacaprioru) [inline drivers/adc/adc_demo/adc_demo.c:156]: should be consistent in returns (return 0 return 1  or return SUCCESS return -ENODEV) 

If there are no more channels you can just return -ENOENT or something like that. There are just a few functions that return something positive like snprintf which returns the length of the string. In this case i would go with typical return SUCCESS or return and error if there is no next ch index but this is not a strong request so i'll leave it up to you.

File: drivers/adc/adc_demo/adc_demo.c [hunk 11]
```diff
@@ -85,6 +112,144 @@ int32_t adc_demo_remove(struct adc_demo_desc *desc)
...
+}
+
+/**
+ * @brief get attributes for adc.
+ * @param device- Physical instance of a iio_demo_device.
+ * @param buf - Where value is stored.
+ * @param len - Maximum length of value to be stored in buf.
+ * @param channel - Channel properties.
+ * @param attr_id - Attribute ID
+ * @return Length of chars written in buf, or negative value on failure.
```
Comment (mirceacaprioru) [inline drivers/adc/adc_demo/adc_demo.c:199]: this is a good example of a positive return related to chars written in a buffer.

File: drivers/adc/adc_demo/adc_demo.h (hunk 1 above)
Comment (mchindri) [reply drivers/adc/adc_demo/adc_demo.h:86]: yes

File: drivers/adc/adc_demo/adc_demo.c (hunk 10 above)
Comment (mchindri) [reply drivers/adc/adc_demo/adc_demo.c:156]: Well, this may be for me, I first wrote this function in iio_demo: 
The reason to return 1 was just to be more human readable in:
while(get_next_ch_idx ...
instead of while (0 == get_next_ch_idx)

Comment (mchindri) [reply drivers/adc/adc_demo/adc_demo.c:156]: any way, I think we can add this to util.h beacause the same logic may be used in new iio drivers

File: drivers/adc/adc_demo/iio_adc_demo.h (hunk 5 above)
Comment (mchindri) [reply drivers/adc/adc_demo/iio_adc_demo.h:54]: prepare_transfer shuld only update the mask with the activ channels, it is how iio works.

When selecting from the oscilloscope channel 0 the mask will be 0x00000001,  when selecting channel 1, it will be 0x00000002 and if both will be 0x00000003

Comment (ranechita) [reply drivers/adc/adc_demo/iio_adc_demo.h]: My bad, I found this out and I also had to modify some other things and now it has update_adc_channels

**Comment by ranechita** [comment]
v6

- re-added [k % ARRAY_SIZE(sine_lut)];
- modified `.prepare_transfer =` to update_adc_channels and update_dac_channels
- modified the init_adc_channels and init_dac_channels to work with osc
- added calls in main for init_adc_channels and init_dac_channels, to initialize the channels before adding the devices
-  moified from `void* dev ` to `struct adc_demo_desc* desc` in init_adc_channels and init_dac_channels

**Comment by ranechita** [comment]
v7

- fixed issues with attributes, in get/set attributes functions and in init_adc/dac_channels functions
- ran astyle.sh

File: drivers/adc/adc_demo/adc_demo.h [hunk 12]
```diff
@@ -45,27 +45,46 @@
...
+#include "iio_types.h"
 
 /******************************************************************************/
 /*************************** Types Declarations *******************************/
 /******************************************************************************/
 
-#define MAX_REG_ADDR		16
+#define MAX_ADC_ADDR		16
+#define DEFAULT_LOCAL_SAMPLES 128
+#define SINE_VECTOR_SIZE 128
```
Comment (mchindri) [inline drivers/adc/adc_demo/adc_demo.h:56]: Not used

File: drivers/adc/adc_demo/adc_demo.h [hunk 13]
```diff
@@ -45,27 +45,46 @@
 /******************************************************************************/
 
 #include <stdint.h>
+#include "iio_types.h"
```
Comment (mchindri) [inline drivers/adc/adc_demo/adc_demo.h:48]: Not needed here

File: drivers/adc/adc_demo/adc_demo.h [hunk 14]
```diff
@@ -45,27 +45,46 @@
...
-	uint8_t reg[MAX_REG_ADDR];
+	uint8_t reg[MAX_ADC_ADDR];
+	/** Demo global device attribute */
+	uint32_t adc_global_attr;
+	/** Demo device channel attribute */
+	uint32_t adc_ch_attr;
+	/** Active channel**/
+	uint32_t active_ch;
+	/** Array of buffers for each channel*/
+	uint16_t **loopback;
```
Comment (mchindri) [inline drivers/adc/adc_demo/adc_demo.h:72]: Maybe you can rename this to to be consistent with the one in init param

File: drivers/dac/dac_demo/dac_demo.c [hunk 15]
```diff
@@ -86,12 +91,164 @@ int32_t dac_demo_remove(struct dac_demo_desc *desc)
...
+	return -EINVAL;
+}
+/**********************************************************************//**
  * @brief read function for the dac demo driver
  * @param desc - descriptor for the dac
  * @param reg_index - the address at which we want to read
  * @param readval- the value read from register
  * @return SUCCESS in case of success, negative error code otherwise.
-*******************************************************************************/
+**************************************************************************/
```
Comment (mchindri) [inline drivers/dac/dac_demo/dac_demo.c:251]: These changes are not needed

File: drivers/adc/adc_demo/iio_adc_demo.h [hunk 16]
```diff
@@ -42,10 +42,45 @@
...
+	.buffer_attributes = NULL,	\
+	.prepare_transfer = update_adc_channels,	\
+	.end_transfer = close_adc_channels,	\
+	.read_dev = (int32_t (*)())adc_read_samples,	\
+	.debug_reg_read = (int32_t (*)()) adc_demo_reg_read,	\
+	.debug_reg_write = (int32_t (*)()) adc_demo_reg_write	\
+}
+
+
+int32_t init_adc_channels(struct adc_demo_desc* dev, uint32_t mask);
```
Comment (mchindri) [inline drivers/adc/adc_demo/iio_adc_demo.h:84]: I would initialize as here https://github.com/analogdevicesinc/no-OS/blob/60c83ad3b11fe54de086746dce6607301399316b/iio/iio_adxrs290/iio_adxrs290.h#L144 for examples.
iio_adc_demo and iio_dac_demo should serve as example for new drivers and be as simple as possible. 
Therefore I think is better to use a fixed number of channels because the chips we are creating drivers for have a fixed number too.

**Comment by ranechita** [comment]
v8

- fixed styling issues
- removed unnecessary includes and defines

File: drivers/adc/adc_demo/iio_adc_demo.h (hunk 16 above)
Comment (dbogdan) [reply drivers/adc/adc_demo/iio_adc_demo.h:83]: Having the option to choose a different number of channels would help us testing the entire infrastructure much easier - we want to make sure everything works even if for example a device has 1 channel or has 16 channels.

**Comment by ranechita** [comment]
v9

- added a channel list of 16 channels
- device number of channels configurable by modifying ADC_CHANNEL_NO/DAC_CHANNEL_NO

File: drivers/adc/adc_demo/adc_demo.c [hunk 17]
```diff
@@ -41,11 +41,34 @@
 /***************************** Include Files **********************************/
 /******************************************************************************/
 
-#include <stdlib.h>
-#include "adc_demo.h"
+#include <stdio.h>
+#include <string.h>
+#include <inttypes.h>
+#include "iio_types.h"
+#include "iio_adc_demo.h"
```
Comment (mchindri) [inline drivers/adc/adc_demo/adc_demo.c:48]: iio_adc_demo.h shouldn't be needed here (the driver doesn't need to know about iio)

File: drivers/adc/adc_demo/adc_demo.c [hunk 18]
```diff
@@ -85,6 +112,163 @@ int32_t adc_demo_remove(struct adc_demo_desc *desc)
...
+
+	if(!dev)
+		return -ENODEV;
+
+	desc = dev;
+
+	if(desc->loopback_buffers == NULL) {
+		//default sin function
+		for(k = 0 ; k < samples; k++)
+			buff[k] = sine_lut[k % ARRAY_SIZE(sine_lut)];
```
Comment (mchindri) [inline drivers/adc/adc_demo/adc_demo.c:195]:
```suggestion
                int offset_per_ch = ARRAY_SIZE(sine_lut) / TOTAL_CHANNELS;
		for(int i = 0; i < samples; i++) {
		     while(get_next_ch_idx(desc->active_ch,ch,&ch))
			buff[k++] = sine_lut[(i + ch * offset_per_ch ) % ARRAY_SIZE(sine_lut)];
```
This will print a sin with an offset for each channel

File: drivers/adc/adc_demo/adc_demo.h [hunk 19]
```diff
@@ -50,22 +50,39 @@
...
  * @brief Desciptor.
  */
 struct adc_demo_desc {
 	/** Dummy registers of device for testing */
-	uint8_t reg[MAX_REG_ADDR];
+	uint8_t reg[MAX_ADC_ADDR];
+	/** Demo global device attribute */
+	uint32_t adc_global_attr;
+	/** Demo device channel attribute */
+	uint32_t adc_ch_attr;
```
Comment (mchindri) [inline drivers/adc/adc_demo/adc_demo.h:66]: uint32_t adc_ch_attr[TOTAL_CHANNELS]

This way you cand change an attribute for each channel.
And at line 263: desc->adc_ch_attr[channel->ch_num] = value;

File: drivers/adc/adc_demo/iio_adc_demo.c [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+*******************************************************************************/
+
+#include <stdio.h>
+#include <string.h>
+#include <inttypes.h>
+#include "iio_adc_demo.h"
+#include "iio_types.h"
+#include "error.h"
+
+/*struct iio_channel iio_adc_channels[MAX_NR_CHANNELS];
```
Comment (mchindri) [inline drivers/adc/adc_demo/iio_adc_demo.c:47]: You commented this by mistake I assume

Comment (ranechita) [reply drivers/adc/adc_demo/iio_adc_demo.c:47]: This file was supposed to be deleted, but i deleted it only form sdk and it remained in the drivers folder

**Comment by ranechita** [comment]
v10

- deleted iio_adc_demo.c and iio_dac_demo.c
- added offset for each channel for sin
- moved `enum iio_adc_demo_attributes { ADC_CHANNEL_ATTR, ADC_GLOBAL_ATTR, };` to adc_demo.h and replaced the include iio_adc_demo.h with adc_demo.h

**Comment by ranechita** [comment]
v11

- added different value for each channel attribute

File: drivers/adc/adc_demo/iio_adc_demo.h [hunk 21]
```diff
@@ -42,10 +42,86 @@
...
+	.debug_attributes = NULL,	\
+	.buffer_attributes = NULL,	\
+	.prepare_transfer = update_adc_channels,	\
+	.end_transfer = close_adc_channels,	\
+	.read_dev = (int32_t (*)())adc_read_samples,	\
+	.debug_reg_read = (int32_t (*)()) adc_demo_reg_read,	\
+	.debug_reg_write = (int32_t (*)()) adc_demo_reg_write	\
+}
+
+//static struct iio_device adc_demo_iio_descriptor = ADC_DEMO_DEV(MODIFIED_CHANNEL_VALUE);
```
Comment (amiclaus) [inline drivers/adc/adc_demo/iio_adc_demo.h:123]: are these 2 commented lines on purpose in this file?

File: drivers/dac/dac_demo/iio_dac_demo.h [hunk 22]
```diff
@@ -42,10 +42,82 @@
...
+	IIO_DEMO_DAC_CHANNEL(9),
+	IIO_DEMO_DAC_CHANNEL(10),
+	IIO_DEMO_DAC_CHANNEL(11),
+	IIO_DEMO_DAC_CHANNEL(12),
+	IIO_DEMO_DAC_CHANNEL(13),
+	IIO_DEMO_DAC_CHANNEL(14),
+	IIO_DEMO_DAC_CHANNEL(15)
+};
+
+//static struct iio_device dac_demo_iio_descriptor = DAC_DEMO_DEV(2);
```
Comment (amiclaus) [inline drivers/dac/dac_demo/iio_dac_demo.h:121]: same mention as above. Is this line needed, since it is commented?

File: projects/iio_demo/src.mk [hunk 23]
```diff
@@ -8,13 +8,26 @@ SRCS +=	$(NO-OS)/util/xml.c						\
 	$(NO-OS)/util/fifo.c						\
 	$(NO-OS)/util/util.c
 
+#drivers
+SRCS += $(DRIVERS)/adc/adc_demo/adc_demo.c				\
+		$(DRIVERS)/dac/dac_demo/dac_demo.c				\
+		$(DRIVERS)/dac/dac_demo/iio_dac_demo.c				\
+		$(DRIVERS)/adc/adc_demo/iio_adc_demo.c				\
+
+
```
Comment (amiclaus) [inline projects/iio_demo/src.mk:17]: nitpick: extra line

File: projects/iio_demo/src.mk [hunk 24]
```diff
@@ -8,13 +8,26 @@ SRCS +=	$(NO-OS)/util/xml.c						\
...
 INCS += $(INCLUDE)/xml.h						\
 	$(INCLUDE)/fifo.h						\
 	$(INCLUDE)/uart.h						\
 	$(INCLUDE)/list.h						\
 	$(INCLUDE)/util.h						\
 	$(INCLUDE)/error.h
 
+
```
Comment (amiclaus) [inline projects/iio_demo/src.mk:25]: nitpick: extra line

File: projects/iio_demo/src.mk [hunk 25]
```diff
@@ -72,4 +85,3 @@ INCS += $(PROJECT)/src/hal/stm32/main.h \
 
 LSCRIPT = $(PROJECT)/src/linker/STM32F446RETX_FLASH.ld
 endif
-
```
Comment (amiclaus) [inline projects/iio_demo/src.mk:75]: nitpick: I think this modification is unrelated to the PR.

File: projects/iio_demo/src/app/main.c [hunk 26]
```diff
@@ -146,6 +149,9 @@ int32_t platform_init()
 	return 0;
 }
 
+//uint16_t buf[2][1000];
```
Comment (amiclaus) [inline projects/iio_demo/src/app/main.c:152]: this can be removed.

File: projects/iio_demo/src/app/main.c [hunk 27]
```diff
@@ -194,12 +212,48 @@ int main(void)
...
+	dac_init_par = (struct dac_demo_init_param) {
+		.loopback_buffers = NULL,
+		.channel_no = 2,
+		.dev_global_attr = 4444,
+		.dev_ch_attr = {1111,1112,1113,1114,1115,1116,1117,1118,1119,1120,1121,1122,1123,1124,1125,1126}
+	};
+	status = dac_demo_init(&dac_desc, &dac_init_par);
+	if (status != SUCCESS)
+		return status;
+	//MODIFIED_CHANNEL_VALUE = 16;
```
Comment (amiclaus) [inline projects/iio_demo/src/app/main.c:244]: is this still required?

File: drivers/adc/adc_demo/adc_demo.c [hunk 28]
```diff
@@ -85,6 +113,166 @@ int32_t adc_demo_remove(struct adc_demo_desc *desc)
...
+	if(!dev)
+		return -ENODEV;
+
+	desc = dev;
+
+	if(desc->loopback_buffers == NULL) {
+		//default sin function
+		int offset_per_ch = ARRAY_SIZE(sine_lut) / TOTAL_CHANNEL_NO;
+		for(int i = 0; i < samples; i++) {
+			while(get_next_ch_idx(desc->active_ch,ch,&ch))
```
Comment (amiclaus) [inline drivers/adc/adc_demo/adc_demo.c:197]: nitpick:
`get_next_ch_idx(desc->active_ch,ch,&ch)` -> `get_next_ch_idx(desc->active_ch, ch, &ch)`
check throughout the entire code.

File: drivers/adc/adc_demo/iio_adc_demo.h [hunk 29]
```diff
@@ -42,10 +42,86 @@
 
 #include <stdlib.h>
 #include "adc_demo.h"
+#include "iio_types.h"
 
-struct iio_device adc_demo_iio_descriptor = {
-	.debug_reg_read = (int32_t (*)()) adc_demo_reg_read,
-	.debug_reg_write = (int32_t (*)()) adc_demo_reg_write
+#define DEFAULT_CHANNEL_NO 2
+#define ADC_CHANNEL_NO changeable_channel_number
```
Comment (mchindri) [inline drivers/adc/adc_demo/iio_adc_demo.h:48]: I would move this to adc_demo.h as:
#infdef MODIFIED_CHANNEL_VALUE
#define ADC_CHANNEL_NO 2
#else
#define ADC_CHANNEL_NO 16
#endif

File: projects/iio_demo/src/app/main.c [hunk 30]
```diff
@@ -146,6 +149,9 @@ int32_t platform_init()
 	return 0;
 }
 
+//uint16_t buf[2][1000];
+uint16_t buf[DEFAULT_CHANNEL_NO][MAX_SAMPLES_PER_CHANNEL];
```
Comment (mchindri) [inline projects/iio_demo/src/app/main.c:153]: use ADC_CHANNEL_NO

File: drivers/adc/adc_demo/adc_demo.c [hunk 31]
```diff
@@ -85,6 +113,166 @@ int32_t adc_demo_remove(struct adc_demo_desc *desc)
...
+				buff[k++] = sine_lut[(i + ch * offset_per_ch ) % ARRAY_SIZE(sine_lut)];
+		}
+
+		return samples;
+	}
+
+	for(int i = 0; i < samples; i++) {
+		while(get_next_ch_idx(desc->active_ch,ch,&ch))
+			buff[k++] = desc->loopback_buffers[ch][i % DEFAULT_LOCAL_SAMPLES];
```
Comment (mchindri) [inline drivers/adc/adc_demo/adc_demo.c:206]: I would define DEFAULT_LOCAL_SAMPLES as a field in init param as: ch_buffer_size

File: drivers/dac/dac_demo/dac_demo.c [hunk 32]
```diff
@@ -37,25 +37,28 @@
  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  *******************************************************************************/
 
-/******************************************************************************/
```
Comment (mchindri) [inline drivers/dac/dac_demo/dac_demo.c:40]: Can you remove comments modifications?

File: drivers/dac/dac_demo/dac_demo.c [hunk 33]
```diff
@@ -86,12 +93,164 @@ int32_t dac_demo_remove(struct dac_demo_desc *desc)
...
+	if(!dev)
+		return -ENODEV;
+
+	desc = dev;
+
+	for(int i = 0; i < samples; i++)
+		while (get_next_ch_idx(desc->active_ch, ch, &ch))
+			desc->loopback_buffers[ch][i % DEFAULT_LOCAL_SAMPLES] = buff[k++];
+
+
```
Comment (mchindri) [inline drivers/dac/dac_demo/dac_demo.c:179]: extra line

File: projects/iio_demo/src/app/main.c [hunk 34]
```diff
@@ -194,12 +212,48 @@ int main(void)
...
+		return status;
+	//MODIFIED_CHANNEL_VALUE = 16;
+	uint32_t adc_channels = ADC_CHANNEL_NO;
+	struct iio_device adc_demo_iio_descriptor = ADC_DEMO_DEV(adc_channels);
+
+	uint32_t dac_channels = DAC_CHANNEL_NO;
+	struct iio_device dac_demo_iio_descriptor = DAC_DEMO_DEV(dac_channels);
+
 	struct iio_app_device devices[] = {
-		IIO_APP_DEVICE("demo_device_output", iio_demo_out_desc,
```
Comment (mchindri) [inline projects/iio_demo/src/app/main.c:198]: When removing this you also have to remove remaining old iio_demo structures.
Anyway, maybe you can just keep old ones in this commit and in an other commit remove this an all related with old iio_demo

File: drivers/adc/adc_demo/iio_adc_demo.h (hunk 21 above)
Comment (ranechita) [reply drivers/adc/adc_demo/iio_adc_demo.h:123]: It is commented on purpose, i left them because i was unsure if the initialization of the number of channels is going to be done through MODIFIED_CHANNEL_VALUE in main or if I should try a different approach

File: drivers/adc/adc_demo/iio_adc_demo.h (hunk 29 above)
Comment (ranechita) [reply drivers/adc/adc_demo/iio_adc_demo.h:48]: I wrote them this way so that it can be used with values other than 2 and 16

**Comment by ranechita** [comment]
v12

- removed extra lines
- corrected `get_next_ch_idx(desc->active_ch,ch,&ch)` to `get_next_ch_idx(desc->active_ch, ch, &ch)`
- replaced DEFAULT_CHANNEL_NO with ADC_CHANNEL_NO
- removed some unnecessary comments

File: drivers/adc/adc_demo/iio_adc_demo.h (hunk 29 above)
Comment (amiclaus) [reply drivers/adc/adc_demo/iio_adc_demo.h:48]: then I think you can do something like this:

**#ifndef ADC_CHANNEL_NO**
#infdef MODIFIED_CHANNEL_VALUE
#define ADC_CHANNEL_NO 2
#else
#define ADC_CHANNEL_NO 16
#endif
**#endif**

This allows you to define custom `ADC_CHANNEL_NO` macro inside your main app directly.

And you can avoid passing static variables to macros, which I don't think is a good practice.

**Review by dbogdan — APPROVED** [review]

**Review by adrimbarean — APPROVED** [review]
